src/query/functions/src/aggregates/adaptors/aggregate_sort_adaptor.rs

```diff
@@ -0,0 +1,275 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::alloc::Layout;
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use borsh::BorshDeserialize;
+use borsh::BorshSerialize;
+use databend_common_column::bitmap::Bitmap;
+use databend_common_exception::Result;
+use databend_common_expression::types::DataType;
+use databend_common_expression::AggrState;
+use databend_common_expression::AggrStateRegistry;
+use databend_common_expression::AggrStateType;
+use databend_common_expression::AggregateFunction;
+use databend_common_expression::AggregateFunctionRef;
+use databend_common_expression::BlockEntry;
+use databend_common_expression::ColumnBuilder;
+use databend_common_expression::DataBlock;
+use databend_common_expression::InputColumns;
+use databend_common_expression::Scalar;
+use databend_common_expression::SortColumnDescription;
+use databend_common_expression::Value;
+use itertools::Itertools;
+
+use crate::aggregates::borsh_deserialize_state;
+use crate::aggregates::borsh_serialize_state;
+use crate::aggregates::AggregateFunctionSortDesc;
+
+#[derive(BorshSerialize, BorshDeserialize, Debug)]
+pub struct SortAggState {
+    columns: Vec<Vec<Scalar>>,
+    data_types: Vec<DataType>,
+}
+
+pub struct AggregateFunctionSortAdaptor {
+    inner: AggregateFunctionRef,
+    argument_len: usize,
+    sort_descs: Vec<AggregateFunctionSortDesc>,
+}
+
+impl AggregateFunction for AggregateFunctionSortAdaptor {
+    fn name(&self) -> &str {
+        self.inner.name()
+    }
+
+    fn return_type(&self) -> Result<DataType> {
+        self.inner.return_type()
+    }
+
+    #[inline]
+    fn init_state(&self, place: AggrState) {
+        Self::set_state(place, SortAggState {
+            columns: vec![],
+            data_types: vec![],
+        });
+        self.inner.init_state(place.remove_first_loc());
+    }
+
+    fn register_state(&self, registry: &mut AggrStateRegistry) {
+        registry.register(AggrStateType::Custom(Layout::new::<SortAggState>()));
+        self.inner.register_state(registry);
+    }
+
+    fn accumulate(
+        &self,
+        place: AggrState,
+        columns: InputColumns,
+        validity: Option<&Bitmap>,
+        input_rows: usize,
+    ) -> Result<()> {
+        let state = Self::get_state(place);
+        Self::init_columns(columns, state, Some(input_rows));
+
+        match validity {
+            Some(validity) => {
+                for (i, column) in columns.iter().enumerate() {
+                    column.iter().zip(validity.iter()).for_each(|(v, b)| {
```

Comment:
If the order by column and the argument column are the same, we don't need to store data for both columns, we need to remove duplicate args in `InputColumns::Block`
for example `select array_agg(a) within group (order by a asc nulls first) from t;`